flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -843,24 +839,138 @@ private boolean cleanUpJobGraph(JobID jobId, boolean cleanupHA) {
+                                completedCheckpointStoreCleanup.cleanupJobData(jobId, jobStatus);
+                                return true;
+                            } catch (Exception e) {
+                                log.warn(
+                                        "An error occurred while cleaning up the CompletedCheckpointStore for job {}.",
+                                        jobId,
+                                        e);
+                                return false;
+                            }
+                        }));
+
+        cleanupTaskResults.add(
+                CompletableFuture.supplyAsync(
+                        () -> cleanupHighAvailabilityServices(jobId), ioExecutor));
+
+        cleanupTaskResults.add(
+                CompletableFuture.supplyAsync(() -> cleanupBlobServer(jobId, true), ioExecutor));
+
+        FutureUtils.combineAll(cleanupTaskResults)
+                .thenAccept(
+                        cleanupResults -> {
+                            boolean cleanupSuccessful =
+                                    cleanupResults.stream().mapToInt(result -> result ? 0 : 1).sum()
+                                            == 0;
+                            if (cleanupSuccessful) {
+                                markJobAsClean(jobId);
+                            }
+                        });
+    }
+
     private void cleanUpRemainingJobData(JobID jobId, boolean jobGraphRemoved) {
         jobManagerMetricGroup.removeJob(jobId);
+
+        boolean remainingJobDataCleaned = jobGraphRemoved;
         if (jobGraphRemoved) {
-            try {
-                highAvailabilityServices.cleanupJobData(jobId);
-            } catch (Exception e) {
-                log.warn(
-                        "Could not properly clean data for job {} stored by ha services", jobId, e);
-            }
+            remainingJobDataCleaned = cleanupHighAvailabilityServices(jobId);
         }
-        blobServer.cleanupJob(jobId, jobGraphRemoved);
 
-        if (jobGraphRemoved) {
-            try {
-                jobResultStore.markResultAsClean(jobId);
-            } catch (IOException e) {
-                log.warn("Could not properly mark job {} result as clean.", jobId, e);
-            }
+        remainingJobDataCleaned &= cleanupBlobServer(jobId, jobGraphRemoved);
+
+        if (remainingJobDataCleaned) {
+            markJobAsClean(jobId);
+        }
+    }
+
+    private boolean cleanupJobGraph(JobID jobId) {
+        try {
+            jobGraphWriter.removeJobGraph(jobId);
+            return true;
+        } catch (Exception e) {
+            log.warn("Could not properly remove job {} from submitted job graph store.", jobId, e);
+            return false;
+        }
+    }
+
+    private boolean cleanupBlobServer(JobID jobId, boolean jobGraphRemoved) {
+        return blobServer.cleanupJob(jobId, jobGraphRemoved);
+    }
+
+    private boolean cleanupHighAvailabilityServices(JobID jobId) {
```

Comment:
Having an interface was my initial approach. I backed off from it because I realized that `FutureUtils`'s retry mechanism only relies on passing in some callback. The interface didn't bring any value - so I left it out for now.